File: android/demo-simple/src/main/java/com/geniusscansdk/simpledemo/MainActivity.java
package com.geniusscansdk.simpledemo;

import static com.geniusscansdk.scanflow.ScanConfiguration.Filter.AUTOMATIC;
import static com.geniusscansdk.scanflow.ScanConfiguration.Filter.AUTOMATIC_BLACK_AND_WHITE;
import static com.geniusscansdk.scanflow.ScanConfiguration.Filter.AUTOMATIC_COLOR;
import static com.geniusscansdk.scanflow.ScanConfiguration.Filter.NONE;
import static com.geniusscansdk.scanflow.ScanConfiguration.Filter.PHOTO;
import static com.geniusscansdk.scanflow.ScanConfiguration.Filter.SOFT_COLOR;
import static com.geniusscansdk.scanflow.ScanConfiguration.Filter.SOFT_GRAYSCALE;
import static com.geniusscansdk.scanflow.ScanConfiguration.Filter.STRONG_MONOCHROME;

import android.app.Activity;
import android.content.Intent;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.TextView;

import androidx.activity.EdgeToEdge;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RawRes;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.content.ContextCompat;
import androidx.core.content.FileProvider;
import androidx.core.graphics.Insets;
import androidx.core.view.OnApplyWindowInsetsListener;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import com.geniusscansdk.core.LicenseException;
import com.geniusscansdk.scanflow.ScanConfiguration;
import com.geniusscansdk.scanflow.ScanFlow;
import com.geniusscansdk.scanflow.ScanResult;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Arrays;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = MainActivity.class.getSimpleName();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);

        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setTitle("GS SDK Simple Demo");
        ViewCompat.setOnApplyWindowInsetsListener(toolbar, new OnApplyWindowInsetsListener() {
            @NonNull
            @Override
            public WindowInsetsCompat onApplyWindowInsets(@NonNull View v, @NonNull WindowInsetsCompat insets) {
                Insets systemInsets = insets.getInsets(WindowInsetsCompat.Type.systemBars());
                v.setPadding(v.getPaddingLeft(), systemInsets.top, v.getPaddingRight(), v.getPaddingBottom());
                return insets;
            }
        });

        // This code shows how to initialize the SDK with a license key.
        // Without a license key, the SDK runs for 60 seconds and then the app needs to be restarted.
        //
        // GeniusScanSDK.setLicenseKey(this, "<Your license key>");

        findViewById(R.id.scan_camera_button).setOnClickListener(view -> scanFromCamera());
        findViewById(R.id.scan_image_button).setOnClickListener(view -> scanFromImage());
        findViewById(R.id.photo_picker_button).setOnClickListener(view -> scanFromPicker());

        TextView appVersionView = findViewById(R.id.app_version);
        appVersionView.setText("v" + BuildConfig.VERSION_NAME + " (" + BuildConfig.VERSION_CODE + ")");
    }

    private ScanConfiguration createBaseConfiguration() {
        ScanConfiguration scanConfiguration = new ScanConfiguration();
        scanConfiguration.multiPage = true;
        scanConfiguration.multiPageFormat = ScanConfiguration.MultiPageFormat.PDF;
        scanConfiguration.pdfPageSize = ScanConfiguration.PdfPageSize.FIT;
        scanConfiguration.pdfMaxScanDimension = 2000;
        scanConfiguration.jpegQuality = 60;
        scanConfiguration.postProcessingActions = ScanConfiguration.Action.ALL;
        scanConfiguration.photoLibraryButtonHidden = false;
        scanConfiguration.flashButtonHidden = false;
        scanConfiguration.defaultFlashMode = ScanConfiguration.FlashMode.AUTO;
        scanConfiguration.backgroundColor = ContextCompat.getColor(this, R.color.md_theme_background);
        scanConfiguration.foregroundColor = ContextCompat.getColor(this, R.color.md_theme_primary);
        scanConfiguration.highlightColor = Color.BLUE;
        scanConfiguration.availableFilters = Arrays.asList(NONE, AUTOMATIC, AUTOMATIC_BLACK_AND_WHITE,
                AUTOMATIC_COLOR, PHOTO, SOFT_GRAYSCALE, SOFT_COLOR, STRONG_MONOCHROME
        );

        ScanConfiguration.OcrConfiguration ocrConfiguration = new ScanConfiguration.OcrConfiguration();
        ocrConfiguration.languages = Arrays.asList("en-US");

        scanConfiguration.ocrConfiguration = ocrConfiguration;

        return scanConfiguration;
    }

    private void scanFromCamera() {
        ScanConfiguration scanConfiguration = createBaseConfiguration();
        scanConfiguration.source = ScanConfiguration.Source.CAMERA;
        startScanning(scanConfiguration);
    }

    private void scanFromImage() {
        ScanConfiguration scanConfiguration = createBaseConfiguration();
        scanConfiguration.source = ScanConfiguration.Source.IMAGE;

        scanConfiguration.sourceImage = new File(getExternalCacheDir(), "temp.jpg");
        copyFileFromResource(R.raw.scan, scanConfiguration.sourceImage);

        startScanning(scanConfiguration);
    }

    private void scanFromPicker() {
        ScanConfiguration scanConfiguration = createBaseConfiguration();
        scanConfiguration.source = ScanConfiguration.Source.GALLERY;
        startScanning(scanConfiguration);
    }

    private void startScanning(ScanConfiguration scanConfiguration) {
        ScanFlow.scanWithConfiguration(MainActivity.this, scanConfiguration);
    }

    private void copyFileFromResource(@RawRes int fileResId, File destinationFile) {
        if (destinationFile.exists()) {
            return;
        }

        byte[] buff = new byte[1024];
        int read;

        try (InputStream in = getResources().openRawResource(fileResId);
             FileOutputStream out = new FileOutputStream(destinationFile)) {
            while ((read = in.read(buff)) > 0) {
                out.write(buff, 0, read);
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (requestCode == ScanFlow.SCAN_REQUEST && resultCode == Activity.RESULT_OK && data != null) {
            try {
                ScanResult result = ScanFlow.getScanResultFromActivityResult(data);
                Uri uri = FileProvider.getUriForFile(MainActivity.this, BuildConfig.APPLICATION_ID + ".fileprovider", result.multiPageDocument);
                Intent intent = new Intent(Intent.ACTION_VIEW, uri);
                intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
                startActivity(intent);
            } catch (LicenseException e) {
                if (e.errorCode == LicenseException.ErrorCode.ExpiredDemo) {
                    new MaterialAlertDialogBuilder(this)
                            .setMessage(e.getMessage())
                            .setPositiveButton("Restart", (dialog, which) -> restartApp())
                            .show();
                } else {
                    // The license key is invalid or expired, either ask the user to update the app or provide a fallback
                    new MaterialAlertDialogBuilder(this)
                            .setMessage("Please update to the latest version.")
                            .setPositiveButton(android.R.string.ok, (dialogInterface, i) -> {})
                            .show();
                }
            } catch (Exception e) {
                Log.e(TAG, "Error during scan flow", e);
                new MaterialAlertDialogBuilder(this)
                        .setMessage("An error occurred: " + e.getMessage())
                        .setPositiveButton(android.R.string.ok, (dialogInterface, i) -> {})
                        .show();
            }
        } else {
            super.onActivityResult(requestCode, resultCode, data);
        }
    }

    private void restartApp() {
        Intent intent = new Intent(this, MainActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        startActivity(intent);
        finish();

        Runtime.getRuntime().exit(0);
    }
}

File: android/demo-simple/src/main/java/com/geniusscansdk/simpledemo/MainActivity.kt
package com.geniusscansdk.simpledemo

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import androidx.activity.enableEdgeToEdge
import androidx.activity.result.ActivityResultLauncher
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.consumeWindowInsets
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.DocumentScanner
import androidx.compose.material.icons.filled.QrCodeScanner
import androidx.compose.material.icons.outlined.DocumentScanner
import androidx.compose.material.icons.outlined.QrCodeScanner
import androidx.compose.material3.CenterAlignedTopAppBar
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.ListItem
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.core.content.ContextCompat
import androidx.core.content.FileProvider
import com.geniusscansdk.core.GeniusScanSDK
import com.geniusscansdk.scanflow.ScanActivity
import com.geniusscansdk.scanflow.ScanFlowConfiguration
import com.geniusscansdk.simpledemo.helpers.ScanHelper
import com.geniusscansdk.simpledemo.helpers.ScanHelper.createBaseOcrConfiguration
import com.geniusscansdk.simpledemo.ui.theme.SimpleDemoTheme

class MainActivity: AppCompatActivity() {

    private lateinit var scanLauncher : ActivityResultLauncher<ScanFlowConfiguration>

    @SuppressLint("SetTextI18n")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        this.enableEdgeToEdge()

        // This code shows how to initialize the SDK with a license key.
        // Without a license key, the SDK runs for 60 seconds and then the app needs to be restarted.
        if (BuildConfig.GSSDK_LICENSE_KEY.isNotEmpty()) {
            GeniusScanSDK.setLicenseKey(this, BuildConfig.GSSDK_LICENSE_KEY, autoRefresh = true)
        }

        setContentView(ComposeView(this).apply {
            setContent {
                SimpleDemoTheme {
                    HomeScreen(
                        defaultScanFlowClick = ::scanFromCamera,
                        customScanFlowClick = {
                            startActivity(Intent(this@MainActivity, CustomScanFlowActivity::class.java))
                        },
                        barcodeFlowClick = {
                            startActivity(Intent(this@MainActivity, BarcodeActivity::class.java))
                        }
                    )
                }
            }
        })

        scanLauncher = registerForActivityResult(ScanActivity.Contract()) { output ->
            ScanHelper.getScanResult(output, this@MainActivity)?.let { scanResult ->
                val uri = FileProvider.getUriForFile(
                    this, BuildConfig.APPLICATION_ID + ".fileprovider",
                    scanResult.multiPageDocument!!
                )
                val resultIntent = Intent(Intent.ACTION_VIEW, uri)
                resultIntent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                startActivity(resultIntent)
            }
        }
    }

    private fun scanFromCamera() {
        val scanConfiguration = createBaseConfiguration()
        scanConfiguration.source = ScanFlowConfiguration.Source.CAMERA
        startScanning(scanConfiguration)
    }

    private fun createBaseConfiguration(): ScanFlowConfiguration {
        val scanConfiguration = ScanFlowConfiguration()
        scanConfiguration.multiPage = true
        scanConfiguration.multiPageFormat = ScanFlowConfiguration.MultiPageFormat.PDF
        scanConfiguration.pdfPageSize = ScanFlowConfiguration.PdfPageSize.FIT
        scanConfiguration.pdfMaxScanDimension = 2000
        scanConfiguration.jpegQuality = 60
        scanConfiguration.postProcessingActions = ScanFlowConfiguration.Action.ALL
        scanConfiguration.photoLibraryButtonHidden = false
        scanConfiguration.flashButtonHidden = false
        scanConfiguration.defaultFlashMode = ScanFlowConfiguration.FlashMode.AUTO
        scanConfiguration.backgroundColor = ContextCompat.getColor(this, R.color.md_theme_background)
        scanConfiguration.foregroundColor = ContextCompat.getColor(this, R.color.md_theme_primary)
        scanConfiguration.highlightColor = Color.BLUE
        scanConfiguration.availableFilters = listOf(
            ScanFlowConfiguration.Filter.NONE,
            ScanFlowConfiguration.Filter.AUTOMATIC,
            ScanFlowConfiguration.Filter.AUTOMATIC_BLACK_AND_WHITE,
            ScanFlowConfiguration.Filter.AUTOMATIC_COLOR,
            ScanFlowConfiguration.Filter.PHOTO,
            ScanFlowConfiguration.Filter.SOFT_GRAYSCALE,
            ScanFlowConfiguration.Filter.SOFT_COLOR,
            ScanFlowConfiguration.Filter.STRONG_MONOCHROME
        )

        scanConfiguration.ocrConfiguration = createBaseOcrConfiguration()

        return scanConfiguration
    }

    private fun startScanning(scanConfiguration: ScanFlowConfiguration) {
        scanLauncher.launch(scanConfiguration)
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun HomeScreen(
    defaultScanFlowClick: () -> Unit,
    customScanFlowClick: () -> Unit,
    barcodeFlowClick: () -> Unit
) {
    val context = LocalContext.current
    Scaffold(
        topBar = { CenterAlignedTopAppBar(title = { Text("GS SDK Simple Demo") }) }
    ) { innerPadding ->
        Box(modifier = Modifier.padding(innerPadding).consumeWindowInsets(innerPadding).fillMaxSize()) {
            Column(
                horizontalAlignment = Alignment.Start
            ) {
                ListItem(
                    headlineContent = { Text(text = stringResource(R.string.custom_scanning_title)) },
                    leadingContent = {
                        Icon(
                            imageVector = Icons.Filled.DocumentScanner,
                            contentDescription = stringResource(R.string.default_scanflow)
                        )
                    },
                    supportingContent = { Text(text = stringResource(R.string.default_scanflow)) },
                    modifier = Modifier.clickable { defaultScanFlowClick() }
                )

                ListItem(
                    headlineContent = { Text(text = stringResource(R.string.custom_scanning_title)) },
                    leadingContent = {
                        Icon(
                            imageVector = Icons.Outlined.DocumentScanner,
                            contentDescription = stringResource(R.string.default_scanflow)
                        )
                    },
                    supportingContent = { Text(text = stringResource(R.string.custom_scanflow)) },
                    modifier = Modifier.clickable { customScanFlowClick() }
                )

                ListItem(
                    headlineContent = { Text(text = stringResource(R.string.structured_data_scanning)) },
                    leadingContent = {
                        Icon(
                            imageVector = Icons.Outlined.QrCodeScanner,
                            contentDescription = stringResource(R.string.default_scanflow)
                        )
                    },
                    modifier = Modifier.clickable {
                        context.startActivity(Intent(context, StructuredDataActivity::class.java))
                    }
                )

                ListItem(
                    headlineContent = { Text(text = stringResource(R.string.barcode_scanning)) },
                    leadingContent = {
                        Icon(
                            imageVector = Icons.Filled.QrCodeScanner,
                            contentDescription = stringResource(R.string.barcode_scanning_description)
                        )
                    },
                    supportingContent = { Text(text = stringResource(R.string.barcode_scanning_description)) },
                    modifier = Modifier.clickable { barcodeFlowClick() }
                )
            }

            Text(
                text = "v${BuildConfig.VERSION_NAME} (${BuildConfig.VERSION_CODE})",
                modifier = Modifier.align(Alignment.BottomCenter).padding(bottom = 10.dp)
            )
        }
    }
}

@Preview
@Composable
private fun HomeScreenPreview() {
    SimpleDemoTheme {
        HomeScreen({}, {}, {})
    }
}
